R/ifelse_censor_linter.R

```diff
@@ -0,0 +1,45 @@
+#' Block usage of ifelse where pmin or pmax is more appropriate
+#'
+#' `ifelse(x > M, M, x)` is the same as `pmin(x, M)`, but harder
+#'   to read and requires several passes over the vector.
+#'
+#' The same goes for other similar ways to censor a vector, e.g.
+#'   `ifelse(x <= M, x, M)` is `pmin(x, M)`,
+#'   `ifelse(x < m, m, x)` is `pmax(x, m)`, and
+#'   `ifelse(x >= m, x, m)` is `pmax(x, m)`.
+#'
+#' @evalRd rd_tags("ifelse_censor_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+ifelse_censor_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    xpath <- glue::glue("//expr[
+      expr[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(ifelse_funs)} ]]
+      and expr[2][
+        (LT or GT or LE or GE)
+        and expr[1] = following-sibling::expr
+        and expr[2] = following-sibling::expr
+      ]
+    ]")
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+
+    return(lapply(
+      bad_expr,
+      xml_nodes_to_lint,
+      source_file = source_file,
+      lint_message = paste(
+        "pmin(x, y) is preferable to ifelse(x < y, x, y);",
+        "pmax(x, y) is preferable to ifelse(x > y, x, y).",
```

Comment:
It would be nice to make the message adapt to the case at hand.
This would especially help here because the difference is hard to visually discern (especially if the argument order is swapped, e.g. `ifelse(x < y, y, x)` needs to be replaced by `pmax(x, y)`). WDYT?

Reply:
I had the same thought but convinced myself it would be a pain to implement. let me take a pass at it with fresh eyes.

Reply:
Not so bad indeed, thanks for pushing!